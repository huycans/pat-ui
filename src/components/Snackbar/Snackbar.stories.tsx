import React, { useState } from 'react';
import { action } from '@storybook/addon-actions';
import Snackbar from './Snackbar';
import Dropdown from '../Dropdown/Dropdown';
import DropdownOption from '../Dropdown/DropdownOption';
import Button from '../Button';
import { SnackbarProps, SnackbarSeverity } from './Snackbar';

export default {
  title: 'Snackbar',
  component: Snackbar,
};

export const DefaultSnackbar = () => {
  const [open, setOpen] = useState(false);
  return (
    <div>
      <Button onClick={() => setOpen(true)}>
        Click here to display the snackbar
      </Button>
      <Snackbar
        open={open}
        onClose={() => setOpen(false)}
        message="Default Snackbar"
      ></Snackbar>
    </div>
  );
};

export const DiffSnackbarType = () => {
  const [open, setOpen] = useState(false);
  const [severity, setSeverity] = useState<SnackbarSeverity>('default');
  const setSnackbar = (severity: SnackbarSeverity) => {
    setOpen(true);
    setSeverity(severity);
  };
  return (
    <div>
      <Button onClick={() => setSnackbar('default')}>
        Click here to display default snackbar
      </Button>
      <Button onClick={() => setSnackbar('error')}>
        Click here to display error snackbar
      </Button>
      <Button onClick={() => setSnackbar('success')}>
        Click here to display success snackbar
      </Button>
      <Button onClick={() => setSnackbar('info')}>
        Click here to display info snackbar
      </Button>
      <Button onClick={() => setSnackbar('warning')}>
        Click here to display warning snackbar
      </Button>
      <Snackbar open={open} onClose={() => setOpen(false)} severity={severity}>
        Hello from Snackbar
      </Snackbar>
    </div>
  );
};

export const DiffPositionButton = () => {
  const [open, setOpen] = useState(false);
  const [horizontal, setHorizontal] = useState('left');
  const [vertical, setVertical] = useState('bottom');
  return (
    <div>
      <label>Choose a vertical position</label>

      <Dropdown
        onChange={(val) => setVertical(val)}
        placeholder="Choose a vertical position"
      >
        <DropdownOption active={true} value="bottom">
          Bottom
        </DropdownOption>
        <DropdownOption active={false} value="top">
          Top
        </DropdownOption>
        <DropdownOption active={false} value="center">
          Center
        </DropdownOption>
      </Dropdown>

      <label>Choose a horizontal position</label>
      <Dropdown
        onChange={(val) => setHorizontal(val)}
        placeholder="Choose a horizontal position"
      >
        <DropdownOption active={true} value="left">
          Left
        </DropdownOption>
        <DropdownOption active={false} value="right">
          Right
        </DropdownOption>
        <DropdownOption active={false} value="center">
          Center
        </DropdownOption>
      </Dropdown>

      <Button onClick={() => setOpen(true)}>
        Click here to display the snackbar
      </Button>
      <Snackbar
        severity="success"
        open={open}
        vertical={vertical}
        horizontal={horizontal}
        onClose={() => setOpen(false)}
      >
        Snackbar at {vertical} {horizontal}
      </Snackbar>
    </div>
  );
};
