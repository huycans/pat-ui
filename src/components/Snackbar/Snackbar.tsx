import React, { FC, useEffect, useState } from 'react';
import { classNames } from '../../utils/classNames';

export type SnackbarSeverity =
  | 'default'
  | 'success'
  | 'warning'
  | 'info'
  | 'error';

export interface SnackbarProps {
  /** set horizontal position (left, center, right) */
  horizontal?: string;
  /** set vertical position (top, center, bottom) */
  vertical?: string;
  /** message to display, this is used only if there're no children in snackbar */
  message?: string;
  /** after this duration, the snackbar will be hidden*/
  autoHideDuration?: number;
  /** manage snackbar open/close state with this prop; should be passed from a parent component */
  open?: boolean;
  /** if open is provided, this function should also be provided from parent component to set open state to false; if open is provided and onClose is empty, the snackbar will not close */
  onClose?: () => void;
  /** set style of snackbar */
  severity?: SnackbarSeverity;
}

/**
 * A Snackbars inform users of a process that an app has performed or will perform.
 *
 * ```js
 * import {Snackbar} from 'pat-ui'
 * ```
 */
export const Snackbar: FC<SnackbarProps> = (props) => {
  const vertical = 'v_' + props.vertical;
  const horizontal = 'h_' + props.horizontal;
  let { open, severity, children, message } = props;
  //default value for onClose: empty function
  const onClose = props.onClose || (() => {});
  const autoHideDuration = props.autoHideDuration;

  //define class names for snackbar
  let classnames = classNames(
    'snackbar',
    vertical,
    horizontal,
    {
      open: open || false,
    },
    'snackbar__' + severity!
  );

  useEffect(() => {
    //close the snackbar after autoHideDuration
    if (open) {
      setTimeout(() => {
        //remove open from classnames
        // classnames = classnames.replace('open', '');
        onClose!();
      }, autoHideDuration);
    }
  }, [props.open, classnames]);

  const closeBtn = props.onClose ? (
    <span className="snackbar__close" onClick={() => onClose!()}></span>
  ) : null;

  //prioritize children first, then message, then empty string
  const content = children || message || '';
  return (
    <div className={classnames}>
      {content} {closeBtn}
    </div>
  );
};

Snackbar.defaultProps = {
  severity: 'default',
  vertical: 'bottom',
  horizontal: 'left',
  open: true,
  autoHideDuration: 2000,
};

export default Snackbar;
