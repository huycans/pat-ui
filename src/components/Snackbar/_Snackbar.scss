@use "sass:color";
@use "sass:selector";

@mixin easeAnimation($time: 1s) {
  transition: all $time ease-in-out;
}

.snackbar {
  $root-snackbar: &;
  z-index: 100;
  position: fixed;
  border-radius: 15px;
  padding: 10px;
  font-weight: $font-weight-bold;
  transform: scale(0);
  @include easeAnimation(0.3s);

  &.open {
    @include easeAnimation(0.3s);
    transform: scale(1);
  }

  //define position
  &.v_bottom {
    bottom: 10%;
  }

  &.v_top {
    top: 10%;
  }

  &.v_center {
    top: 50%;
  }

  &.h_left {
    left: 10%;
  }

  &.h_center {
    left: 50%;
  }

  &.h_right {
    right: 10%;
  }

  //define background color and text color of snackbar
  @mixin snackbarColor($bgcolor: $white) {
    background-color: $bgcolor;
    color: setTextColor($bgcolor);
  }

  @function setTextColor($bgcolor: $white) {
    //adjust text color based on background color
    // @error '$bgcolor: #{$bgcolor}';
    @if (lightness($bgcolor) > 50) {
      // @error 'return black #{$bgcolor}';
      @return $black;
    } @else {
      // @error 'return white #{$bgcolor}';
      @return $white;
    }
  }

  //define colors based on type of snackbar
  &__default {
    @include snackbarColor;
  }

  &__error {
    @include snackbarColor($danger);
  }

  &__warning {
    @include snackbarColor($warning);
  }

  &__success {
    @include snackbarColor($success);
  }

  &__info {
    @include snackbarColor($info);
  }
  //close button
  // @at-root &__close {
  &__close {
    cursor: pointer;
    margin-left: 20px;
    margin-right: 20px;

    &::after {
      //set background color of this pseudo element to the color of the text
      background-color: currentColor;
      content: '';
      position: absolute;
      // top: 8px;
      width: 18px;
      height: 3px;
      top: 22px;
      right: 10px;
      transform: rotate(45deg);
      @include easeAnimation;
    }

    &::before {
      //set background color of this pseudo element to the color of the text
      background-color: currentColor;
      content: '';
      position: absolute;
      // top: 8px;
      width: 18px;
      height: 3px;
      top: 22px;
      right: 10px;
      transform: rotate(-45deg);
      @include easeAnimation;
    }
  }
}
