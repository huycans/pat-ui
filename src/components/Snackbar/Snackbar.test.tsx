import React from 'react';

import { render, fireEvent, cleanup } from '@testing-library/react';
import Snackbar, { SnackbarProps } from './Snackbar';

//global variable to use
let open: SnackbarProps['open'];
let openSnackbar: Function;
let closeSnackbar: SnackbarProps['onClose'];
let snackbarProps: SnackbarProps;

//function to conver hex color to rgb
function hexToRgbA(hex: string) {
  var c: any;
  if (/^#([A-Fa-f0-9]{3}){1,2}$/.test(hex)) {
    c = hex.substring(1).split('');
    if (c.length == 3) {
      c = [c[0], c[0], c[1], c[1], c[2], c[2]];
    }
    c = '0x' + c.join('');
    return 'rgba(' + [(c >> 16) & 255, (c >> 8) & 255, c & 255].join(',') + ')';
  }
  throw new Error('Bad Hex');
}

beforeAll(() => {
  cleanup();
  jest.clearAllMocks();
  open = true;
  openSnackbar = () => (open = true);
  snackbarProps = {
    open: open,
  };
});

describe('Snackbar', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(
      <Snackbar open={true}> Snapshot Button </Snackbar>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render default snackbar', () => {
    const wrapper = render(
      <Snackbar {...snackbarProps}>Default Snackbar</Snackbar>
    );
    const snackbar = wrapper.queryByText('Default Snackbar') as HTMLElement;
    expect(snackbar).toBeInTheDocument();
    expect(snackbar.tagName).toBe('DIV');
    expect(snackbar).toHaveClass('snackbar v_bottom h_left snackbar__default');
  });

  it('should render different types of snackbar: info', () => {
    //render info
    snackbarProps.severity = 'info';
    let wrapper = render(<Snackbar {...snackbarProps}>Snackbar</Snackbar>);
    let snackbar = wrapper.queryByText('Snackbar') as HTMLElement;
    expect(snackbar).toBeInTheDocument();
    expect(snackbar.tagName).toBe('DIV');
    expect(snackbar).toHaveClass('snackbar v_bottom h_left snackbar__info');
    expect(snackbar).toHaveStyle(`background-color: ${hexToRgbA('#17a2b8')}`);
  });

  it('should render different types of snackbar: error', () => {
    //render error
    snackbarProps.severity = 'error';
    let wrapper = render(<Snackbar {...snackbarProps}>Snackbar</Snackbar>);
    let snackbar = wrapper.queryByText('Snackbar') as HTMLElement;
    expect(snackbar).toBeInTheDocument();
    expect(snackbar.tagName).toBe('DIV');
    expect(snackbar).toHaveClass('snackbar v_bottom h_left snackbar__error');
    expect(snackbar).toHaveStyle(`background-color: ${hexToRgbA('#dc3545')}`);
  });

  it('should render different types of snackbar: success', () => {
    //render success
    snackbarProps.severity = 'success';
    let wrapper = render(<Snackbar {...snackbarProps}>Snackbar</Snackbar>);
    let snackbar = wrapper.queryByText('Snackbar') as HTMLElement;
    expect(snackbar).toBeInTheDocument();
    expect(snackbar.tagName).toBe('DIV');
    expect(snackbar).toHaveClass('snackbar v_bottom h_left snackbar__success');
    expect(snackbar).toHaveStyle(`background-color: ${hexToRgbA('#52c41a')}`);
  });

  it('should render different types of snackbar: warning', () => {
    //render warning
    snackbarProps.severity = 'warning';
    let wrapper = render(<Snackbar {...snackbarProps}>Snackbar</Snackbar>);
    let snackbar = wrapper.queryByText('Snackbar') as HTMLElement;
    expect(snackbar).toBeInTheDocument();
    expect(snackbar.tagName).toBe('DIV');
    expect(snackbar).toHaveClass('snackbar v_bottom h_left snackbar__warning');
    expect(snackbar).toHaveStyle(`background-color: ${hexToRgbA('#fadb14')}`);
  });

  it('should render message if no children is provided', () => {
    snackbarProps.message = 'Snackbar Message';
    const wrapper = render(<Snackbar {...snackbarProps}></Snackbar>);
    const snackbar = wrapper.queryByText('Snackbar Message') as HTMLElement;
    expect(snackbar).toBeInTheDocument();
  });

  //future tests:
  //close the snackbar after autoHideDuration
});
