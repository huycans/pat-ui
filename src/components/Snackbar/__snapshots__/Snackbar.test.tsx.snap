// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snackbar should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="snackbar v_bottom h_left open snackbar__default"
  >
     Snapshot Button  
  </div>
</DocumentFragment>
`;
